tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,318 @@
+use http::Request;
+use std::{
+    task::{Context, Poll},
+    time::Duration,
+};
+use tower_service::Service;
+use tracing::{debug, warn};
+
+pub(crate) struct Timeout<S> {
+    inner: S,
+    server_timeout: Option<Duration>,
+}
+
+impl<S> Timeout<S> {
+    pub(crate) fn new(inner: S, server_timeout: Option<Duration>) -> Self {
+        Self {
+            inner,
+            server_timeout,
+        }
+    }
+}
+
+impl<S, ReqBody> Service<Request<ReqBody>> for Timeout<S>
+where
+    S: Service<Request<ReqBody>>,
+    S::Error: Into<crate::Error>,
+{
+    type Response = S::Response;
+    type Error = crate::Error;
+    type Future = future::TimeoutFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        // Try to parse the `grpc-timeout` header, if it is present
+        let header_timeout = headers::try_parse_grpc_timeout(req.headers()).unwrap_or_else(|e| {
+            warn!("Error parsing grpc-timeout header {:?}", e);
```

Comment:
nit: not sure `warn` is the best fit here since clients might be sending garbage data. It isn't really tonics fault, modulo parsing bugs of course 😊 I kinda think we should use `trace` or drop the log entirely.